[query] corrupted zstd block

### What happened?

https://hail.zulipchat.com/#narrow/stream/223457-Hail-Batch-support/topic/QoB.20Error.3A.20GoogleJsonResponseException.3A.20404.20Not.20Found/near/398355473

> I was running hl.pca on the wheel you created for me -> 0.2.124-fcaafc533ec1. and there seems to be a transient error going on https://batch.hail.is/batches/8069235?q=state+%3D+failed, not sure whether this is the same as the previous ones. I just cancelled the job before error summary appears. and here is the code I am running:



```python3
        vat_ht = hl.read_table(get_aou_util_path(name="vat"))
        vat_ht = vat_ht.collect_by_key()
        meta_ht = hl.read_table(get_sample_meta_path(annotation=True))
        meta_ht = meta_ht.filter(~meta_ht.related)
        pops = args.pops.split(",") if (args.pops is not None) else POPS
        for pop in pops:
            mt = get_filtered_mt(analysis_type='variant', filter_variants=True, filter_samples=False,
                                 adj_filter=True, pop=pop)
            variants_to_keep = vat_ht.filter(
                (vat_ht.locus.in_autosome()) &
                (hl.is_snp(vat_ht.alleles[0], vat_ht.alleles[1])) &
                (vat_ht['values'][f'gvs_{pop}_af'][0] >= 0.0001) &
                ((vat_ht.values[f"gvs_{pop}_an"][0] >= (N_SAMPLES[pop] * 2 * MIN_CALL_RATE[pop])))
            )
            print('Filtering Variants...')
            mt = mt.filter_rows(hl.is_defined(variants_to_keep[mt.row_key])) # filter to high quality variants
            print('Filtering Samples...')
            mt = mt.filter_cols(hl.is_defined(meta_ht[mt.col_key])) # filter to unrelated samples -> later to project
            print('Running PCA...')
            eigenvalues, scores, loadings = hl.pca(
                hl.int(hl.is_defined(mt.GT)),
                compute_loadings=True,
                k=50,
            )
            print('Writing tables...')
            eigenvalues.write(
                get_pca_ht_path(pop=pop, name='evals'),
                overwrite=args.overwrite,
            )
            scores.write(
                get_pca_ht_path(pop=pop, name='scores'),
                overwrite=args.overwrite,
            )
            loadings.write(
                get_pca_ht_path(pop=pop, name='loadings'),
                overwrite=args.overwrite,
            )
```

### Version

0.2.126

### Relevant log output

_No response_

#### Summary

It has happened twice. The failing partition is different in each run.

1. 49340 https://batch.hail.is/batches/8069235/jobs/51280
2. 25997 https://batch.hail.is/batches/8083195/jobs/27937

The pipeline runs two table collects to get sample information, then converts the matrix table to a table of ndarrays of the value `hl.int(hl.is_defined(mt.GT))`.

The entries are getting subsetted, so there is skipping going on.  In both cases, we are decoding the entry array when the corrupted block is discovered. In the first case, we are skipping an int (must be RGQ based on the etype and type). In the second case, we are decoding a string (must be FT).

Since the error happens on a seemingly arbitrary partition, it seems likely this is related to our transient error handling. Both runs use a version of Hail after we fixed the broken transient error handling in GoogleStorageFS (run 1 used fcaafc533e, run 2 used 0.2.126 / ee77707f4f).

---

#### Path forward

If it *is* a transient error, we need to fix how we handle transient errors. Maybe our position handling logic is wrong? If it is *not* a transient error, maybe our skipping logic is wrong? FT appears immediately after RGQ and we know RGQ is getting skipped.

Our implementation of `seek` for the compressed block buffers looks sketchy to me, but we're using PartitionNativeReader which does no seeking.

Action items:
1. Log every transient error.
2. Log the file name and the offset on failure.

---

#### Debugging information

EType:
```
+EBaseStruct{
  `the entries! [877f12a8827e18f61222c6c8c5fb04a8]`:+EArray[
    EBaseStruct{
      GT:EInt32,
      GQ:EInt32,
      RGQ:EInt32,
      FT:EBinary,
      AD:EArray[EInt32]}]}
```
(zipped) Type:
```
Struct{
  locus:Locus(GRCh38),
  alleles:Array[String],
  filters:Set[String],
  info:Struct{
    AC:Array[Int32]},
  `the entries! [877f12a8827e18f61222c6c8c5fb04a8]`:Array[
    Struct{
      GT:Call,
      GQ:Int32,
      FT:String,
      AD:Array[Int32]}]}
```
Source buffer spec:
```
{"name":"LEB128BufferSpec","child":
  {"name":"BlockingBufferSpec","blockSize":65536,"child":
    {"name":"ZstdBlockBufferSpec","blockSize":65536,"child":
      {"name":"StreamBlockBufferSpec"}}}}
```
Error for run 1.
```
Caused by: com.github.luben.zstd.ZstdException: Corrupted block detected
	at com.github.luben.zstd.ZstdDecompressCtx.decompressByteArray(ZstdDecompressCtx.java:157) ~[zstd-jni-1.5.2-1.jar:1.5.2-1]
	at is.hail.io.ZstdInputBlockBuffer.readBlock(InputBuffers.scala:655) ~[gs:__hail-query-ger0g_jars_fcaafc533ec130ae210b152afa81c0b5ac04592b.jar.jar:0.0.1-SNAPSHOT]
	at is.hail.io.BlockingInputBuffer.ensure(InputBuffers.scala:385) ~[gs:__hail-query-ger0g_jars_fcaafc533ec130ae210b152afa81c0b5ac04592b.jar.jar:0.0.1-SNAPSHOT]
	at is.hail.io.BlockingInputBuffer.readByte(InputBuffers.scala:403) ~[gs:__hail-query-ger0g_jars_fcaafc533ec130ae210b152afa81c0b5ac04592b.jar.jar:0.0.1-SNAPSHOT]
	at is.hail.io.LEB128InputBuffer.readByte(InputBuffers.scala:220) ~[gs:__hail-query-ger0g_jars_fcaafc533ec130ae210b152afa81c0b5ac04592b.jar.jar:0.0.1-SNAPSHOT]
	at is.hail.io.LEB128InputBuffer.skipInt(InputBuffers.scala:260) ~[gs:__hail-query-ger0g_jars_fcaafc533ec130ae210b152afa81c0b5ac04592b.jar.jar:0.0.1-SNAPSHOT]
	at __C796collect_distributed_array_table_native_writer.__m872SKIP_o_int32(Unknown Source) ~[?:?]
	at __C796collect_distributed_array_table_native_writer.__m869INPLACE_DECODE_o_struct_of_o_int32ANDo_int32ANDo_int32ANDo_binaryANDo_array_of_o_int32END_TO_o_struct_of_o_callANDo_int32ANDo_stringANDo_array_of_o_int32END(Unknown Source) ~[?:?]
	at __C796collect_distributed_array_table_native_writer.__m868INPLACE_DECODE_r_array_of_o_struct_of_o_int32ANDo_int32ANDo_int32ANDo_binaryANDo_array_of_o_int32END_TO_r_array_of_o_struct_of_o_callANDo_int32ANDo_stringANDo_array_of_o_int32END(Unknown Source) ~[?:?]
	at __C796collect_distributed_array_table_native_writer.__m867DECODE_r_struct_of_r_array_of_o_struct_of_o_int32ANDo_int32ANDo_int32ANDo_binaryANDo_array_of_o_int32ENDEND_TO_SBaseStructPointer(Unknown Source) ~[?:?]
```

Error for run 2
```
Caused by: com.github.luben.zstd.ZstdException: Corrupted block detected
	at com.github.luben.zstd.ZstdDecompressCtx.decompressByteArray(ZstdDecompressCtx.java:157) ~[zstd-jni-1.5.2-1.jar:1.5.2-1]
	at is.hail.io.ZstdInputBlockBuffer.readBlock(InputBuffers.scala:655) ~[gs:__hail-query-ger0g_jars_ee77707f4fab22b1c253321b082a70aff3f44d1c.jar.jar:0.0.1-SNAPSHOT]
	at is.hail.io.BlockingInputBuffer.readBytes(InputBuffers.scala:444) ~[gs:__hail-query-ger0g_jars_ee77707f4fab22b1c253321b082a70aff3f44d1c.jar.jar:0.0.1-SNAPSHOT]
	at is.hail.io.LEB128InputBuffer.readBytes(InputBuffers.scala:253) ~[gs:__hail-query-ger0g_jars_ee77707f4fab22b1c253321b082a70aff3f44d1c.jar.jar:0.0.1-SNAPSHOT]
	at __C816collect_distributed_array_table_native_writer.__m893INPLACE_DECODE_o_binary_TO_o_string(Unknown Source) ~[?:?]
	at __C816collect_distributed_array_table_native_writer.__m889INPLACE_DECODE_o_struct_of_o_int32ANDo_int32ANDo_int32ANDo_binaryANDo_array_of_o_int32END_TO_o_struct_of_o_callANDo_int32ANDo_stringANDo_array_of_o_int32END(Unknown Source) ~[?:?]
	at __C816collect_distributed_array_table_native_writer.__m888INPLACE_DECODE_r_array_of_o_struct_of_o_int32ANDo_int32ANDo_int32ANDo_binaryANDo_array_of_o_int32END_TO_r_array_of_o_struct_of_o_callANDo_int32ANDo_stringANDo_array_of_o_int32END(Unknown Source) ~[?:?]
	at __C816collect_distributed_array_table_native_writer.__m887DECODE_r_struct_of_r_array_of_o_struct_of_o_int32ANDo_int32ANDo_int32ANDo_binaryANDo_array_of_o_int32ENDEND_TO_SBaseStructPointer(Unknown Source) ~[?:?]
```

Still replicates using Wenhan's pipeline: https://batch.hail.is/batches/8090848/jobs/21993

I confirmed that it basically failed immediately after a transient error occurred:
```
2023-11-30 20:56:12.836 services: WARN: Encountered 1 transient errors, most recent one was com.github.luben.zstd.ZstdException: Corrupted block detected. Delay is 1781.
2023-11-30 20:56:12.865 : INFO: TaskReport: stage=0, partition=20053, attempt=0, peakBytes=61343744, peakBytesReadable=58.50 MiB, chunks requested=1630, cache hits=1618
2023-11-30 20:56:12.866 : INFO: RegionPool: FREE: 58.5M allocated (10.5M blocks / 48.0M chunks), regions.size = 21, 0 current java objects, thread 9: pool-2-thread-1
2023-11-30 20:56:12.871 WorkerTimer$: INFO: executeFunction took 18072.278761 ms.
2023-11-30 20:56:12.871 GoogleStorageFS$: INFO: createNoCompression: gs://aou_tmp/parallelizeAndComputeWithIndex/qFyj_OTLPksF7rgiwE62GTDErymYe4DI3ENyNJbpe2s=/result.20053
2023-11-30 20:56:12.938 GoogleStorageFS$: INFO: close: gs://aou_tmp/parallelizeAndComputeWithIndex/qFyj_OTLPksF7rgiwE62GTDErymYe4DI3ENyNJbpe2s=/result.20053
2023-11-30 20:56:13.036 GoogleStorageFS$: INFO: closed: gs://aou_tmp/parallelizeAndComputeWithIndex/qFyj_OTLPksF7rgiwE62GTDErymYe4DI3ENyNJbpe2s=/result.20053
2023-11-30 20:56:13.037 WorkerTimer$: INFO: writeOutputs took 165.518872 ms.
2023-11-30 20:56:13.037 Worker$: INFO: finished job 20053 at root gs://aou_tmp/parallelizeAndComputeWithIndex/qFyj_OTLPksF7rgiwE62GTDErymYe4DI3ENyNJbpe2s=
2023-11-30 20:56:13.037 Worker$: INFO: throwing the exception so that this Worker job is marked as failed.
2023-11-30 20:56:13.037 JVMEntryway: ERROR: QoB Job threw an exception.
java.lang.reflect.InvocationTargetException: null
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_382]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_382]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_382]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_382]
	at is.hail.JVMEntryway$1.run(JVMEntryway.java:119) ~[jvm-entryway.jar:?]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) ~[?:1.8.0_382]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) ~[?:1.8.0_382]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) ~[?:1.8.0_382]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) ~[?:1.8.0_382]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) ~[?:1.8.0_382]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) ~[?:1.8.0_382]
	at java.lang.Thread.run(Thread.java:750) ~[?:1.8.0_382]
Caused by: com.github.luben.zstd.ZstdException: Corrupted block detected
	at com.github.luben.zstd.ZstdDecompressCtx.decompressByteArray(ZstdDecompressCtx.java:157) ~[zstd-jni-1.5.2-1.jar:1.5.2-1]
	at is.hail.io.ZstdInputBlockBuffer.readBlock(InputBuffers.scala:671) ~[gs:__hail-query-ger0g_jars_dking_mn15emaev5kt_04fe3bfc818e56afd6cb922299a4612565a2146c.jar.jar:0.0.1-SNAPSHOT]
	at is.hail.io.BlockingInputBuffer.ensure(InputBuffers.scala:393) ~[gs:__hail-query-ger0g_jars_dking_mn15emaev5kt_04fe3bfc818e56afd6cb922299a4612565a2146c.jar.jar:0.0.1-SNAPSHOT]
	at is.hail.io.BlockingInputBuffer.readByte(InputBuffers.scala:411) ~[gs:__hail-query-ger0g_jars_dking_mn15emaev5kt_04fe3bfc818e56afd6cb922299a4612565a2146c.jar.jar:0.0.1-SNAPSHOT]
	at is.hail.io.LEB128InputBuffer.readByte(InputBuffers.scala:226) ~[gs:__hail-query-ger0g_jars_dking_mn15emaev5kt_04fe3bfc818e56afd6cb922299a4612565a2146c.jar.jar:0.0.1-SNAPSHOT]
	at is.hail.io.LEB128InputBuffer.readInt(InputBuffers.scala:232) ~[gs:__hail-query-ger0g_jars_dking_mn15emaev5kt_04fe3bfc818e56afd6cb922299a4612565a2146c.jar.jar:0.0.1-SNAPSHOT]
	at __C816collect_distributed_array_table_native_writer.__m890INPLACE_DECODE_o_int32_TO_o_call(Unknown Source) ~[?:?]
	at __C816collect_distributed_array_table_native_writer.__m889INPLACE_DECODE_o_struct_of_o_int32ANDo_int32ANDo_int32ANDo_binaryANDo_array_of_o_int32END_TO_o_struct_of_o_callANDo_int32ANDo_stringANDo_array_of_o_int32END(Unknown Source) ~[?:?]
	at __C816collect_distributed_array_table_native_writer.__m888INPLACE_DECODE_r_array_of_o_struct_of_o_int32ANDo_int32ANDo_int32ANDo_binaryANDo_array_of_o_int32END_TO_r_array_of_o_struct_of_o_callANDo_int32ANDo_stringANDo_array_of_o_int32END(Unknown Source) ~[?:?]
	at __C816collect_distributed_array_table_native_writer.__m887DECODE_r_struct_of_r_array_of_o_struct_of_o_int32ANDo_int32ANDo_int32ANDo_binaryANDo_array_of_o_int32ENDEND_TO_SBaseStructPointer(Unknown Source) ~[?:?]
```

--- a/hail/build.gradle
+++ b/hail/build.gradle
@@ -168,7 +168,7 @@ dependencies {
         throw new UnsupportedOperationException("Couldn't pick a valid elasticsearch.")
     }
 
-    implementation(group: 'com.google.cloud', name: 'google-cloud-storage', version: '2.29.1') {
+    implementation(group: 'com.google.cloud', name: 'google-cloud-storage', version: '2.30.1') {
         exclude group: 'com.fasterxml.jackson.core'
     }
 

